# Eigenvector of gene regulatory networks and gene expression proﬁles

One of the principal steps towards an understanding of bacterial gene regulation and therefore cellular function is to quantitatively assess, how well a given `gene regulatory network` explains given `gene expression data` set (i.e., the activity proﬁle of all/many genes). An interesting mathematical approach for addressing this question is to compare **`spectral` properties of the gene regulatory network** (in particular, the `eigenvectors` of the graph) with the **activity patterns of genes** (which are vectors, where each node of the network is characterized by a real number). 



Gene expression data will be taken from the GEO database. The gene regulatory network will be downloaded from RegulonDB.



**keywords**:

spectral graph theory, gene regulatory network, gene expression data, eigenvector, Escherichia coli



Further reading:

Clough, E., and Barrett, T. (2016). The gene expression omnibus database. Statistical Genomics: Methods and Protocols, 93-110.

Gama-Castro, S., Salgado, H., Santos-Zavaleta, A., Ledezma-Tejeida, D., Muniz-Rascado, L., Garcia-Sotelo, J. S., ... and Medina-Rivera, A. (2015). RegulonDB version 9.0: high-level integration of gene regulation, coexpression, motif clustering and beyond. Nucleic Acids Research, 44(D1), D133-D143.